vllm/model_executor/models/molmo.py

```diff
@@ -0,0 +1,1130 @@
+        self.multimodal_config = multimodal_config
+
+        vision_config = VisionBackboneConfig()
+        self.vision_backbone = MolmoVisionBackbone(config, vision_config, quant_config)
+        self.model = MolmoModel(config, cache_config, quant_config)
+
+        if self.config.weight_tying:
+            self.lm_head = self.model.transformer.wte
+        else:
+            self.lm_head = ParallelLMHead(
+                config.embedding_size or config.vocab_size,
+                config.hidden_size,
+                quant_config=quant_config,
+            )
+
+        self.logits_processor = LogitsProcessor(config.embedding_size or config.vocab_size)
+        self.sampler = Sampler()
+        
+    def forward(
+        self,
+        input_ids: torch.LongTensor,
+        positions: torch.LongTensor,
+        kv_caches: List[torch.Tensor],
+        attn_metadata: AttentionMetadata,
+        images: Optional[torch.Tensor] = None,
+        image_masks: Optional[torch.Tensor] = None,
+        image_input_idx: Optional[torch.Tensor] = None,
+        seq_len: Optional[torch.Tensor] = None,
+        **kwargs,
+    ) -> SamplerOutput:
+
+        has_image = images is not None
+        num_image: Optional[int] = None
+
+        if has_image:
+            x = self.model.embed_tokens(input_ids)
+            batch_size = images.size(0)
+            if not isinstance(seq_len, torch.Tensor):
+                seq_len = torch.tensor(seq_len, device=x.device)
+            images = images.to(device=x.device, dtype=x.dtype)
+            image_features: torch.Tensor
+            image_features, cls_embed = self.vision_backbone(images=images, image_masks=image_masks)
+            num_image, num_patch = image_features.shape[1:3]
+            assert image_input_idx.shape == (batch_size, num_image, num_patch)
+
+            image_features = image_features.to(x.device)
+
+            # insert the image feature into the embedding.
+            image_features = image_features.view(batch_size, num_image * num_patch, -1)
+            image_input_idx = image_input_idx.view(batch_size, num_image * num_patch)
+
+            valid = image_input_idx >= 0
+            image_features = image_features * valid[:, :, None].to(image_features.dtype)
+            image_features = image_features.view(batch_size * num_image * num_patch, -1).contiguous()
+
+            image_input_idx = image_input_idx * valid.to(image_input_idx.dtype)
+            offset = torch.cat([seq_len.new_zeros((1)), seq_len.cumsum(dim=0)[:-1]], dim=0)[:, None]
+            image_input_idx = image_input_idx + offset.to(image_input_idx.dtype)
+            image_input_idx = image_input_idx.flatten()[:, None]
+            mat = image_input_idx == torch.arange(seq_len.sum().item(), device=x.device)[None, :]
+            mat = mat.to(image_features.dtype)
+            x = x + torch.einsum('nd,nm->md', image_features, mat)
+
+            inputs_embeds = x
+
+            input_ids = None
```

Comment:
Can you rewrite this section so that we can follow the same code pattern as below as in other LMMs?

```
        image_input = self._parse_and_validate_image_input(**kwargs)

        if image_input is not None:
            vision_embeddings = self._process_image_input(image_input)
            inputs_embeds = self.language_model.model.get_input_embeddings(
                input_ids)

            inputs_embeds = merge_multimodal_embeddings(
                input_ids, inputs_embeds, vision_embeddings,
                self.vision_args.image_token_id)

            input_ids = None
        else:
            inputs_embeds = None

        hidden_states = self.language_model.model(input_ids,
                                                  positions,
                                                  kv_caches,
                                                  attn_metadata,
                                                  None,
                                                  inputs_embeds=inputs_embeds)
```
Feel free to check out other model files (e.g., `llava-next.py`)

Reply:
Hi, we can rewrite the code, but I have a question: is this just for making code more readable?

Reply:
@sangho-vision Mostly yes - in vLLM we care a lot about coding standard and the way to define the inference pattern in the forward pass, especially for multimodal models since it can get complicated. 

I'm happy to help too!